pkg/controller/istiocsr/controller.go

```diff
@@ -0,0 +1,355 @@
+		Watches(&rbacv1.ClusterRoleBinding{}, handler.EnqueueRequestsFromMapFunc(mapFunc), controllerManagedResourcePredicates).
+		Watches(&rbacv1.Role{}, handler.EnqueueRequestsFromMapFunc(mapFunc), controllerManagedResourcePredicates).
+		Watches(&rbacv1.RoleBinding{}, handler.EnqueueRequestsFromMapFunc(mapFunc), controllerManagedResourcePredicates).
+		Watches(&corev1.Service{}, handler.EnqueueRequestsFromMapFunc(mapFunc), controllerManagedResourcePredicates).
+		Watches(&corev1.ServiceAccount{}, handler.EnqueueRequestsFromMapFunc(mapFunc), controllerManagedResourcePredicates).
+		WatchesMetadata(&corev1.Secret{}, handler.EnqueueRequestsFromMapFunc(mapFunc), controllerWatchResourcePredicates).
+		Complete(r)
+}
+
+// Reconcile function to compare the state specified by the IstioCSR object against the actual cluster state,
+// and to make the cluster state reflect the state specified by the user.
+func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	r.log.V(1).Info("reconciling", "request", req)
+
+	// Fetch the istiocsr.openshift.operator.io CR
+	istiocsr := &v1alpha1.IstioCSR{}
+	if err := r.Get(ctx, req.NamespacedName, istiocsr); err != nil {
+		if errors.IsNotFound(err) {
+			// NotFound errors, since they can't be fixed by an immediate
+			// requeue (have to wait for a new notification), and can be processed
+			// on deleted requests.
+			r.log.V(1).Info("istiocsr.openshift.operator.io object not found, skipping reconciliation", "request", req)
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, fmt.Errorf("failed to fetch istiocsr.openshift.operator.io %q during reconciliation: %w", req.NamespacedName, err)
+	}
+
+	if !istiocsr.DeletionTimestamp.IsZero() {
+		r.log.V(1).Info("istiocsr.openshift.operator.io is marked for deletion", "namespace", req.NamespacedName)
+
+		if requeue, err := r.cleanUp(istiocsr); err != nil {
+			return ctrl.Result{}, fmt.Errorf("clean up failed for %q istiocsr.openshift.operator.io instance deletion: %w", req.NamespacedName, err)
+		} else if requeue {
+			return ctrl.Result{RequeueAfter: defaultRequeueTime}, nil
+		}
+
+		if err := r.removeFinalizer(ctx, istiocsr, finalizer); err != nil {
+			return ctrl.Result{}, err
+		}
+
+		r.log.V(1).Info("removed finalizer, cleanup complete", "request", req.NamespacedName)
+		return ctrl.Result{}, nil
+	}
+
+	// Set finalizers on the istiocsr.openshift.operator.io resource
+	if err := r.addFinalizer(ctx, istiocsr); err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to update %q istiocsr.openshift.operator.io with finalizers: %w", req.NamespacedName, err)
+	}
+
+	return r.processReconcileRequest(istiocsr, req.NamespacedName)
+
+}
+
+func (r *Reconciler) processReconcileRequest(istiocsr *v1alpha1.IstioCSR, req types.NamespacedName) (res ctrl.Result, err error) {
+	statusUpdateRequired := false
+	defer func() {
+		// using named return values to handle status update in deferred function
+		if statusUpdateRequired {
```

Comment:
Defer can be avoided as it is a conditional code execution and does not have common execution required at the exit of the function.

Following function calls can be put togther in a different function and a call made in the present place where the flag is set.
istiocsr.Status.SetCondition()
statusUpdateRequired = true
r.updateStatus(r.ctx, istiocsr)

Reply:
Yeah, added it for better code readability, instead of calling same set of code or a method, would mean checking the error and setting the Result. Please let me know if it needs to be updated as suggested.

Reply:
> instead of calling same set of code

Clubbing it to a common function helps to achieve it. It is easy to read in the sequence of execution than to refer to boolean variable usage. It'd be best to incorporate the suggestions as post the status update, there is no re-iteration in the implementation that'd recur the function call.

Reply:
suggestion incorporated.